File: packages/cli/src/targets/vscode/actions/run-command.js
// @ts-check
/// <reference path="./globals.js" />

/**
 * @param {{ command?: string, shell?: 'cmd' | 'powershell' | 'bash' | 'sh' | 'zsh' | 'fish' } | null} payload
 * @param {(result: { command: string, stdout: string, stderr: string, exitCode: number }) => void} reply
 * @param {import('vscode').OutputChannel} channel
 * @returns {void}
 */
function runCommand(payload, reply, channel) {
  const command = payload?.command ?? ''
  const isWin = os.platform() === 'win32'
  const requestedShell = payload?.shell
  const [shell, flag] = resolveShell(requestedShell, isWin)
  execFile(shell, [flag, command], { timeout: 30000 }, (err, stdout, stderr) => {
    reply({ command, stdout, stderr, exitCode: err?.code ?? 0 })
  })
}

/**
 * @param {string | undefined} requested
 * @param {boolean} isWin
 * @returns {[string, string]}
 */
function resolveShell(requested, isWin) {
  switch (requested) {
    case 'cmd':
      return ['cmd', '/c']
    case 'powershell':
      return ['powershell', '-Command']
    case 'bash':
      return ['bash', '-c']
    case 'sh':
      return ['sh', '-c']
    case 'zsh':
      return ['zsh', '-c']
    case 'fish':
      return ['fish', '-c']
    default:
      return isWin ? ['cmd', '/c'] : ['sh', '-c']
  }
}

File: packages/cli/src/targets/jetbrains/actions/RunCommand.kt
fun handleRunCommand(payload: org.json.JSONObject?, reply: org.json.JSONObject) {
    if (payload == null) return
    val command = payload.optString("command", "")
    val requestedShell = payload.optString("shell", "")
    val isWindows = System.getProperty("os.name").lowercase().contains("win")
    val (shell, flag) = resolveShell(requestedShell, isWindows)
    val result = runShellCommand(command, shell, flag)
    reply.put("type", "run-command:reply")
    val replyPayload = org.json.JSONObject()
    replyPayload.put("command", command)
    replyPayload.put("stdout", result.first)
    replyPayload.put("stderr", result.second)
    replyPayload.put("exitCode", result.third)
    reply.put("payload", replyPayload)
}

fun resolveShell(requested: String, isWindows: Boolean): Pair<String, String> = when (requested) {
    "cmd"        -> Pair("cmd", "/c")
    "powershell" -> Pair("powershell", "-Command")
    "bash"       -> Pair("bash", "-c")
    "sh"         -> Pair("sh", "-c")
    "zsh"        -> Pair("zsh", "-c")
    "fish"       -> Pair("fish", "-c")
    else         -> if (isWindows) Pair("cmd", "/c") else Pair("sh", "-c")
}
